src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -208,30 +232,62 @@ CoreWorkerDirectTaskSubmitter::GetOrConnectLeaseClient(
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
     const SchedulingKey &scheduling_key, const rpc::Address *raylet_address) {
-  if (pending_lease_requests_.find(scheduling_key) != pending_lease_requests_.end()) {
+  auto &scheduling_key_entry = scheduling_key_entries_[scheduling_key];
+  auto &pending_lease_request = scheduling_key_entry.pending_lease_request_;
+
+  if (pending_lease_request.first) {
     // There's already an outstanding lease request for this type of task.
     return;
   }
-  auto it = task_queues_.find(scheduling_key);
-  if (it == task_queues_.end()) {
+
+  auto &task_queue = scheduling_key_entry.task_queue_;
+  if (task_queue.empty()) {
     // We don't have any of this type of task to run.
     return;
   }
 
+  // If pipelining is enabled, check whether we really need a new worer or whether we have
+  // enough room in an existing worker's pipeline to send the new tasks
+  if (max_tasks_in_flight_per_worker_ > 1) {
+    if (scheduling_key_entry.tot_tasks_in_flight <
+        scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
+      // The pipelines to the current workers are not full yet, so we don't need more
+      // workers.
+
+      // Find a worker with a number of tasks in flight that is less than the maximum
+      // value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle to send tasks to
+      // that worker
+      for (auto active_worker_addr : scheduling_key_entry.active_workers_) {
+        RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
+                  worker_to_lease_entry_.end());
+        auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
+        if (lease_entry.tasks_in_flight_ < max_tasks_in_flight_per_worker_) {
+          OnWorkerIdle(active_worker_addr, scheduling_key, false,
+                       lease_entry.assigned_resources_);
+          break;
```

Comment:
I'm not sure if we want to break here. Couldn't there be multiple idle workers that could get filled by new tasks? But I guess this depends on where we decide to put this logic (see above comment).

Reply:
I see! So I guess we would remove the break statement if we put this in `SubmitTask`, right?

Reply:
Hmm I need to think about that. It seems like we could structure the code so that we're always guaranteed that only one worker needs to be filled up during `SubmitTask`, but I'm not sure.